### ISSUE TITLE
QSys Interrupts, Component Metadata

### ISSUE THREAD

**Issue by cutephoton** [SpinalHDL/SpinalHDL#172]
Have some proposed changes I'd like some feedback on.

A new tag to add metadata to components:
```
case class ComponentMetaTag (
    displayName     : String        = null,
    author          : String        = null,
    version         : String        = null,
    category        : String        = null,
    description     : String        = null
) extends SpinalTag
```

The purpose of this change is to improve QSys generation.

In addition, changing the QSysify interrupt support to include senders and receivers (in the same manner avalon MM makes slaves and masters). Add the following and update interrupt generation code as needed.

```
case class InterruptTag(clockDomain : ClockDomain, addressablePoint : Data = null) extends SpinalTag
```
I swapped clockDomain and addressablePoint (and left the original class in place). The purpose of addressablePoint doesn't make much sense to me besides causing QSys to complain. It is optional.

**edit** As well, should add to conduit:
```
case class ConduitSignalTag(signalType : String = "export") extends SpinalTag
```
Conduits can also represent interfaces between components if the conduits contain the same set of signal types and appropriate signal directions. The export signal type can only be exported out of the QSys design.

**Comment by Dolu1990** [comment]
Sorry, was quite busy those last days XD

So, yes, +1 for the component tags.
About QSys interrupt, i think the reason why they are currently not linked to a given memory bus, is because that's the most general way of describing them.

I never used interruption linked to avalon busses, so, hard for me to say if that's a good thing or not^^
So that's good as you keep it optional :)

**Comment by cutephoton** [comment]
No problem. I apologize for the length! ;) As I've poured over the present and past issues, I can't help but notice that you have a good track record of being helpful, engaged and responsive. (Doesn't hurt that it's one of the few projects that has Avalon and QSys integration...)

**What component metadata should be included?**

The reason I ask is that the structure as defined today maps 1 to 1 basically to QSys. The category field (group in QSys) is a hierarchical path, but is that a good idea? Or should be a QSys specific field.

Some fields I didn't include that may be worth including:

- Company
- Github/Project URL
- License?
- Documentation references (QSys supports links to files and URLs)

**What values should be used by Spinal's built in modules?**

I'd like to update the QSys code for some of the built in modules. Here's an example for Interrupt controller. It is defined in the Altera code only, but if it's generic could be applied to the component itself.

```
toplevel addTag(ComponentMetaTag(
        displayName     = "SpinalHDL Interrupt Controller (Avalon)",
        author          = "SpinalHDL",
        //version         = "1.0", --> Use Spinal Version Number
        category        = "SpinalHDL/Peripherals",
        description     = "This module multiplexes multiple IRQ senders on to a single IRQ."
    ))
```
![image](https://user-images.githubusercontent.com/19512121/50717876-9036b800-103f-11e9-9860-0ad4af4359af.png)

![image](https://user-images.githubusercontent.com/19512121/50717887-a5134b80-103f-11e9-8b9d-88b05692b151.png)

**Related: Is it feasible to get a memory map for a peripheral's register file?**

It would be useful to be able to extract information about peripherals/slaves and such. Mostly interested in the memory map, but even more interesting if one can identify the fields. Eventually, would use to generate headers, documentation, etc.

**Other QSys Changes to support code gen**

I made some additional changes that allow module/interface "assignments" (key/value pairs). I can use this later to generate code for the vex platform. Intel also associates some useful metadata to its peripherals. Reading Intel/Altera's docs, the intention was to enable code generation. I'm doing a bit of work in automating certain system s/w platform stuff like memory maps.

I'll submit a pull request at some point. I'm going to hold off for now... I want to see how the code gen work above goes before finalizing this work.

It's not necessary to look at the stuff below but if you're interested in what I've been able to do with the debuginfo file I've included it here. And now it's time to generate some useful artifacts. :)

```
[qsys   INFO    ] ---- Memory Maps ----
[qsys   INFO    ] Bus base.dbus
[qsys   INFO    ]     ddr_0_mem_if_ddr3_emif.avl                         0x20000000 0x40000000
[qsys   INFO    ]     mcu_ram0.s1                                        0x80000000 0x80008000
[qsys   INFO    ]     ls_0_periph_id.control_slave                       0xF0000000 0xF0000008
[qsys   INFO    ]     ls_0_AvalonInterruptCtrl_0.io_bus                  0xF0000010 0xF0000018
[qsys   INFO    ]     ls_0_periph_timer_systick.s1                       0xF0000100 0xF0000120
[qsys   INFO    ]     ls_0_periph_timer_1.s1                             0xF0000120 0xF0000140
[qsys   INFO    ]     ls_0_periph_gpio_led.s1                            0xF0000200 0xF0000210
[qsys   INFO    ]     ls_0_periph_gpio.s1                                0xF0000220 0xF0000230
[qsys   INFO    ]     ls_0_i2c_0.csr                                     0xF0000300 0xF0000340
[qsys   INFO    ]     ls_0_AvalonMMUartCtrl_0.io_bus                     0xF0000400 0xF0000410
[qsys   INFO    ] 
[qsys   INFO    ] Bus base.ibus
[qsys   INFO    ]     ddr_0_mem_if_ddr3_emif.avl                         0x20000000 0x40000000
[qsys   INFO    ]     mcu_ram0.s1                                        0x80000000 0x80008000
[qsys   INFO    ] 
[qsys   INFO    ] 
[qsys   INFO    ] ---- IRQ List ----
[qsys   INFO    ]     base.irqexternal                                   irq_mapper.sender                                  0
[qsys   INFO    ]     base.irqtimer                                      irq_mapper_001.sender                              0
[qsys   INFO    ]     base_VexRiscvAvalon_0.timerInterrupt               irq_mapper_001.sender                              0
[qsys   INFO    ]     base_VexRiscvAvalon_0.externalInterrupt            irq_mapper.sender                                  0
[qsys   INFO    ]     ls_0_AvalonInterruptCtrl_0.io_irqReceiver          ls_0_i2c_0.interrupt_sender                        0
[qsys   INFO    ]     ls_0_AvalonInterruptCtrl_0.io_irqReceiver          ls_0_AvalonMMUartCtrl_0.io_interrupt               3
[qsys   INFO    ]     ls_0_AvalonInterruptCtrl_0.io_irqReceiver          ls_0_periph_gpio.irq                               2
[qsys   INFO    ]     ls_0_AvalonInterruptCtrl_0.io_irqReceiver          ls_0_periph_timer_1.irq                            1

```

```
[xo     INFO    ]     Module (Tag: module) {
[xo     INFO    ]       name = 'ls_0_AvalonInterruptCtrl_0'
[xo     INFO    ]       kind = 'AvalonInterruptCtrl'
[xo     INFO    ]       parameters = [
[xo     INFO    ]         Parameter (Tag: parameter) {
[xo     INFO    ]           name = 'AUTO_IO_IRQRECEIVER_INTERRUPTS_USED'
[xo     INFO    ]           value = '15'
[xo     INFO    ]           type = 'java.math.BigInteger'
[xo     INFO    ]         }
[xo     INFO    ]         Parameter (Tag: parameter) {
[xo     INFO    ]           name = 'deviceFamily'
[xo     INFO    ]           value = 'UNKNOWN'
[xo     INFO    ]           type = 'java.lang.String'
[xo     INFO    ]         }
[xo     INFO    ]         Parameter (Tag: parameter) {
[xo     INFO    ]           name = 'generateLegacySim'
[xo     INFO    ]           value = 'false'
[xo     INFO    ]           type = 'boolean'
[xo     INFO    ]         }
[xo     INFO    ]       ]
[xo     INFO    ]       interfaces = [
[xo     INFO    ]         Interface (Tag: interface) {
[xo     INFO    ]           name = 'io_bus'
[xo     INFO    ]           kind = 'avalon_slave'
[xo     INFO    ]           assignments = [
[xo     INFO    ]             Assignment (Tag: assignment) {
[xo     INFO    ]               name = 'embeddedsw.configuration.isFlash'
[xo     INFO    ]               value = '0'
[xo     INFO    ]             }
[xo     INFO    ]             Assignment (Tag: assignment) {
[xo     INFO    ]               name = 'embeddedsw.configuration.isMemoryDevice'
[xo     INFO    ]               value = '0'
[xo     INFO    ]             }
[xo     INFO    ]             Assignment (Tag: assignment) {
[xo     INFO    ]               name = 'embeddedsw.configuration.isNonVolatileStorage'
[xo     INFO    ]               value = '0'
[xo     INFO    ]             }
[xo     INFO    ]             Assignment (Tag: assignment) {
[xo     INFO    ]               name = 'embeddedsw.configuration.isPrintableDevice'
[xo     INFO    ]               value = '0'
[xo     INFO    ]             }
[xo     INFO    ]           ]
[xo     INFO    ]           interrupt = [
[xo     INFO    ]           ]
[xo     INFO    ]           ports = [
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_bus_address'
[xo     INFO    ]               direction = 'Input'
[xo     INFO    ]               width = '3'
[xo     INFO    ]               role = 'address'
[xo     INFO    ]             }
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_bus_askRead'
[xo     INFO    ]               direction = 'Input'
[xo     INFO    ]               width = '1'
[xo     INFO    ]               role = 'read'
[xo     INFO    ]             }
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_bus_askWrite'
[xo     INFO    ]               direction = 'Input'
[xo     INFO    ]               width = '1'
[xo     INFO    ]               role = 'write'
[xo     INFO    ]             }
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_bus_waitRequestn'
[xo     INFO    ]               direction = 'Output'
[xo     INFO    ]               width = '1'
[xo     INFO    ]               role = 'waitrequest_n'
[xo     INFO    ]             }
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_bus_writeData'
[xo     INFO    ]               direction = 'Input'
[xo     INFO    ]               width = '32'
[xo     INFO    ]               role = 'writedata'
[xo     INFO    ]             }
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_bus_readDataValid'
[xo     INFO    ]               direction = 'Output'
[xo     INFO    ]               width = '1'
[xo     INFO    ]               role = 'readdatavalid'
[xo     INFO    ]             }
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_bus_readData'
[xo     INFO    ]               direction = 'Output'
[xo     INFO    ]               width = '32'
[xo     INFO    ]               role = 'readdata'
[xo     INFO    ]             }
[xo     INFO    ]           ]
[xo     INFO    ]           memorymap = <NoneType: omitted>
[xo     INFO    ]         }
[xo     INFO    ]         Interface (Tag: interface) {
[xo     INFO    ]           name = 'io_irqReceiver'
[xo     INFO    ]           kind = 'interrupt_receiver'
[xo     INFO    ]           assignments = [
[xo     INFO    ]           ]
[xo     INFO    ]           interrupt = [
[xo     INFO    ]             Interrupt (Tag: interrupt) {
[xo     INFO    ]               fromSlaveModule = 'ls_0_i2c_0'
[xo     INFO    ]               fromSlaveInterface = 'interrupt_sender'
[xo     INFO    ]               fromName = 'ls_0_i2c_0.interrupt_sender'
[xo     INFO    ]               interfaceName = 'io_irqReceiver'
[xo     INFO    ]               moduleName = 'ls_0_AvalonInterruptCtrl_0'
[xo     INFO    ]               interruptNumber = '0'
[xo     INFO    ]             }
[xo     INFO    ]             Interrupt (Tag: interrupt) {
[xo     INFO    ]               fromSlaveModule = 'ls_0_AvalonMMUartCtrl_0'
[xo     INFO    ]               fromSlaveInterface = 'io_interrupt'
[xo     INFO    ]               fromName = 'ls_0_AvalonMMUartCtrl_0.io_interrupt'
[xo     INFO    ]               interfaceName = 'io_irqReceiver'
[xo     INFO    ]               moduleName = 'ls_0_AvalonInterruptCtrl_0'
[xo     INFO    ]               interruptNumber = '3'
[xo     INFO    ]             }
[xo     INFO    ]             Interrupt (Tag: interrupt) {
[xo     INFO    ]               fromSlaveModule = 'ls_0_periph_gpio'
[xo     INFO    ]               fromSlaveInterface = 'irq'
[xo     INFO    ]               fromName = 'ls_0_periph_gpio.irq'
[xo     INFO    ]               interfaceName = 'io_irqReceiver'
[xo     INFO    ]               moduleName = 'ls_0_AvalonInterruptCtrl_0'
[xo     INFO    ]               interruptNumber = '2'
[xo     INFO    ]             }
[xo     INFO    ]             Interrupt (Tag: interrupt) {
[xo     INFO    ]               fromSlaveModule = 'ls_0_periph_timer_1'
[xo     INFO    ]               fromSlaveInterface = 'irq'
[xo     INFO    ]               fromName = 'ls_0_periph_timer_1.irq'
[xo     INFO    ]               interfaceName = 'io_irqReceiver'
[xo     INFO    ]               moduleName = 'ls_0_AvalonInterruptCtrl_0'
[xo     INFO    ]               interruptNumber = '1'
[xo     INFO    ]             }
[xo     INFO    ]           ]
[xo     INFO    ]           ports = [
[xo     INFO    ]             Port (Tag: port) {
[xo     INFO    ]               name = 'io_irqReceiver'
[xo     INFO    ]               direction = 'Input'
[xo     INFO    ]               width = '32'
[xo     INFO    ]               role = 'irq'
[xo     INFO    ]             }
[xo     INFO    ]           ]
[xo     INFO    ]           memorymap = <NoneType: omitted>
[xo     INFO    ]         }

```

**Comment by cutephoton** [comment]
Thought I should add an update.

I briefly surveyed various documents and standards like IP-Xact to see what's common. There are differences. Didn't see anything major. At the same time, I was exploring how to construct a sufficiently detailed model of a design which would entail other types of metadata. The dependency on QSys to build the Avalon interconnect, however, makes life very very very difficult. I got it to work it was a tightly coupled unmaintainable hot mess held together by duct tape that I decided I needed to rethink my approach.... for the 3rd time. lol. 

To avoid creating churn, I thought it would be prudent to defer until I finish it, give up, or another users asks for it. ;)
